File: sdk/python/ratio/grpc.py
"""The gRPC transport: the same seven calls over ``grpcio``, with the wire
format from :mod:`ratio._wire` — so ``pip install 'ratio-sdk[grpc]'`` is the
whole setup, and there is no stub to regenerate when the contract moves."""

from __future__ import annotations

from typing import Callable

from . import _wire
from .errors import RatioError
from .types import Account, AccountPage, Posting, Transaction, TransactionPage, TrialBalance


class GrpcTransport:
    def __init__(self, endpoint: str, timeout: float = 30.0) -> None:
        try:
            import grpc  # noqa: F401 — optional; REST needs nothing
        except ImportError:
            raise ImportError(
                "the gRPC transport needs grpcio — pip install 'ratio-sdk[grpc]' — "
                "or use transport='rest', which needs nothing"
            ) from None
        self._grpc = grpc
        self._timeout = timeout
        if endpoint.startswith("https://"):
            self._channel = grpc.secure_channel(endpoint[len("https://"):], grpc.ssl_channel_credentials())
        else:
            target = endpoint[len("http://"):] if endpoint.startswith("http://") else endpoint
            self._channel = grpc.insecure_channel(target)

    def _unary(self, service: str, method: str, request: bytes, decode: Callable[[bytes], object]) -> object:
        # No serializers: bytes in, bytes out. The wire module is the codec.
        call = self._channel.unary_unary(f"/ratio.v1.{service}/{method}")
        try:
            return decode(call(request, timeout=self._timeout))
        except self._grpc.RpcError as e:
            raise RatioError(e.code().value[0], e.details() or "") from None

    # ── the seven calls ──────────────────────────────────────────────────

    def get_transaction(self, name: str) -> Transaction:
        return self._unary("Ledger", "GetTransaction", _wire.get_request(name), _wire.decode_transaction)  # type: ignore[return-value]

    def list_transactions(self, parent: str, page_size: int, page_token: str) -> TransactionPage:
        return self._unary(
            "Ledger", "ListTransactions", _wire.list_request(parent, page_size, page_token), _wire.decode_transaction_page
        )  # type: ignore[return-value]

    def create_transaction(
        self, parent: str, postings: list[Posting], transaction_id: str, control_plane_hash: str
    ) -> Transaction:
        t = Transaction(postings=tuple(postings), control_plane_hash=control_plane_hash)
        return self._unary(
            "Ledger", "CreateTransaction", _wire.create_transaction_request(parent, t, transaction_id), _wire.decode_transaction
        )  # type: ignore[return-value]

    def get_account(self, name: str) -> Account:
        return self._unary("Chart", "GetAccount", _wire.get_request(name), _wire.decode_account)  # type: ignore[return-value]

    def list_accounts(self, parent: str, page_size: int, page_token: str) -> AccountPage:
        return self._unary(
            "Chart", "ListAccounts", _wire.list_request(parent, page_size, page_token), _wire.decode_account_page
        )  # type: ignore[return-value]

    def create_account(self, parent: str, account: Account, account_id: str) -> Account:
        return self._unary(
            "Chart", "CreateAccount", _wire.create_account_request(parent, account, account_id), _wire.decode_account
        )  # type: ignore[return-value]

    def get_trial_balance(self, name: str) -> TrialBalance:
        return self._unary("Chart", "GetTrialBalance", _wire.get_request(name), _wire.decode_trial_balance)  # type: ignore[return-value]

    def close(self) -> None:
        self._channel.close()

File: sdk/typescript/src/grpc.ts
import { RatioError } from "./errors.js";
import type { Transport } from "./transport.js";
import type { Account, ListAccountsResponse, ListTransactionsResponse, Posting, Transaction, TrialBalance } from "./types.js";
import * as wire from "./wire.js";

/**
 * The gRPC transport: the same seven calls over `@grpc/grpc-js`, with the wire
 * format from `wire.ts` — so `npm install @grpc/grpc-js` is the whole setup and
 * there is no stub to regenerate when the contract moves.
 */
/** The slice of `grpc.Client` this transport uses. */
interface ClientLike {
  makeUnaryRequest(
    method: string,
    serialize: (v: Uint8Array) => Buffer,
    deserialize: (b: Buffer) => Uint8Array,
    request: Uint8Array,
    callback: (err: { code: number; details: string } | null, value?: Uint8Array) => void,
  ): unknown;
  close(): void;
}

export class GrpcTransport implements Transport {
  private constructor(private readonly client: ClientLike) {}

  /** Connect to `endpoint` (`http://host:port` is plaintext, `https://` is TLS). */
  static async connect(endpoint: string): Promise<GrpcTransport> {
    let grpc: typeof import("@grpc/grpc-js");
    try {
      grpc = await import("@grpc/grpc-js");
    } catch {
      throw new Error("the gRPC transport needs @grpc/grpc-js — npm install @grpc/grpc-js — or use the REST transport, which needs nothing");
    }
    const tls = endpoint.startsWith("https://");
    const target = endpoint.replace(/^https?:\/\//, "");
    const creds = tls ? grpc.credentials.createSsl() : grpc.credentials.createInsecure();
    const client = new grpc.Client(target, creds);
    return new GrpcTransport(client as unknown as ClientLike);
  }

  private unary<T>(service: string, method: string, request: Uint8Array, decode: (b: Uint8Array) => T): Promise<T> {
    return new Promise((resolve, reject) => {
      this.client.makeUnaryRequest(
        `/ratio.v1.${service}/${method}`,
        (v) => Buffer.from(v),
        (b) => new Uint8Array(b),
        request,
        (err, value) => {
          if (err) reject(new RatioError(err.code, err.details ?? ""));
          else resolve(decode(value ?? new Uint8Array()));
        },
      );
    });
  }

  getTransaction(name: string): Promise<Transaction> {
    return this.unary("Ledger", "GetTransaction", wire.encodeGetRequest(name), wire.decodeTransaction);
  }
  listTransactions(parent: string, pageSize: number, pageToken: string): Promise<ListTransactionsResponse> {
    return this.unary("Ledger", "ListTransactions", wire.encodeListRequest(parent, pageSize, pageToken), wire.decodeListTransactionsResponse);
  }
  createTransaction(parent: string, postings: Posting[], transactionId: string, controlPlaneHash: string): Promise<Transaction> {
    const t: Transaction = { postings };
    if (controlPlaneHash) t.controlPlaneHash = controlPlaneHash;
    return this.unary("Ledger", "CreateTransaction", wire.encodeCreateTransactionRequest(parent, t, transactionId), wire.decodeTransaction);
  }
  getAccount(name: string): Promise<Account> {
    return this.unary("Chart", "GetAccount", wire.encodeGetRequest(name), wire.decodeAccount);
  }
  listAccounts(parent: string, pageSize: number, pageToken: string): Promise<ListAccountsResponse> {
    return this.unary("Chart", "ListAccounts", wire.encodeListRequest(parent, pageSize, pageToken), wire.decodeListAccountsResponse);
  }
  createAccount(parent: string, account: Account, accountId: string): Promise<Account> {
    return this.unary("Chart", "CreateAccount", wire.encodeCreateAccountRequest(parent, account, accountId), wire.decodeAccount);
  }
  getTrialBalance(name: string): Promise<TrialBalance> {
    return this.unary("Chart", "GetTrialBalance", wire.encodeGetRequest(name), wire.decodeTrialBalance);
  }
  close(): void {
    this.client.close();
  }
}
